in_toto/verifylib.py

```diff
@@ -634,6 +637,59 @@ def verify_all_item_rules(items, links, target_links=None):
     verify_item_rules(item.name, item.product_matchrules,
         link.products, target_links)
 
+def verify_threshold_equality(layout, step_link_dict):
+  """
+  <Purpose>
+    Iteratively verifies threshold conditions and link artifacts of
+    passed dictionary (step_link_dict).
+
+  <Arguments>
+    layout:
+            The layout specified by the project owner against which the
+            threshold will be verified.
+
+    step_link_dict:
+            A dictionary of key-link pair with step names as keys. For each
+            step name, there are one or more keyids and corresponding
+            link objects.
+
+  <Exceptions>
+    raises an Exception if threshold is not verified.
+    ThresholdVerificationError()
+
+  <Side Effects>
+    None.
+  
+  <Returns>
+    A dictionary containing one Link metadata object per step only if
+    the link artifacts of all link objects are identical for a step.
+  """
+  target_links = {}
+  for key_link in step_link_dict:
+    key_link_dict = step_link_dict[key_link]
+    
+    for step in layout.steps:
+       if step.name is key_link:
+        threshold = step.threshold
+
+    if len(key_link_dict) >= threshold:
+      link_dict = {}
+      if len(key_link_dict) == 1:
+        for keyid in key_link_dict:
+          link_dict = key_link_dict[keyid]
+      else:
+        i = 0
+        for keyid in key_link_dict:
+          if i == 0:
+            link_dict = key_link_dict[keyid]
+            i = i + 1
+          else:
+            if link_dict.materials != key_link_dict[keyid].materials or link_dict.products != key_link_dict[keyid].products:
+              raise ThresholdVerificationError("Link artifacts do not match!".format())
+      target_links[key_link] = link_dict
+    else:
+      raise ThresholdVerificationError("Step not performed by enough functionaries!".format())
```

Comment:
The algorithm is a bit simplified now. Less nesting.
We are skipping over the non-threshold steps.
The function also checks if the link file was indeed signed by different authorized functionaries (or rather was not signed by the same functionary to just satisfy the threshold).